### ISSUE TITLE
Is CommonIntegrateNoseHooverStepKernel::forcesAreValid always false?

### ISSUE THREAD

**Issue by zhi-wang** [openmm/openmm#4715]
Hi Peter,

We've observed that `NoseHooverIntegrator` consistently takes approximately twice as long to run compared to other integrators (e.g., `LangevinMiddleIntegrator`). This behavior persists across different force fields. This issue appears to have been previously mentioned:

> https://github.com/openmm/openmm/issues/4280#issuecomment-1778439277
> "I get [...] via NoseHooverIntegrator , but this takes me more than twice as long."

I noticed potentially relevant code in `CommonKernels.cpp`. The `forcesAreValid` variable appears to lack a corresponding `true` assignment.
> https://github.com/openmm/openmm/blob/53770948682c40bd460b39830d4e0f0fd3a4b868/platforms/common/src/CommonKernels.cpp#L5907

A 2020 PR by Andy Simmonett shows this variable being set to `true`, but this code is no longer present in the current version.

> https://github.com/openmm/openmm/pull/2527/files#diff-77f522a025f8bb35151d2b13dc6f9d6c6b5642acdc8cda996f8bcbacefa66a1dR5783

This is my initial analysis and may be incomplete or incorrect. Any help would be appreciated.

Best,

Zhi Wang

**Comment by peastman** [comment]
It's a good question.  And we should also ask, should that variable exist at all?  Most integrators don't have it.  For example, here's the `step()` method in LangevinMiddleIntegrator:

https://github.com/openmm/openmm/blob/53770948682c40bd460b39830d4e0f0fd3a4b868/openmmapi/src/LangevinMiddleIntegrator.cpp#L90-L98

It calls `calcForcesAndEnergy()` immediately before calling `execute()` on the kernel.  Forces are always valid at the start of `execute()`, because they've just been calculated, and they're never valid at the end, because it changes the particle positions.  This is generally true for leapfrog integrators, but not for velocity Verlet style ones.  Those ones begin with a half step velocity update based on the same forces used at the end of the previous step.  In that case, it's worth tracking the validity of the forces.

Andy's original version of NoseHooverIntegrator used a velocity Verlet discretization, but he converted it to LF-Middle in #2600.  I think that means the `forcesAreValid` parameter is now obsolete and could be removed.

**Comment by wenyan4work** [comment]
How should we fix it?

**Comment by peastman** [comment]
Thanks for bumping this.  I just posted a PR with the fix.

**Comment by z-gong** [comment]
In PR #4753, the following lines were removed from `void NoseHooverIntegrator::step(int steps)` 
```
        if(context->updateContextState())
            forcesAreValid = false;
```
Could this removal prevent force objects that update the context before integration (e.g., barostats) from functioning correctly?

**Comment by peastman** [comment]
Oops!  Thanks for catching that.  It's fixed in #4754.
